testutil/network/gatewaynet/network.go

```diff
@@ -0,0 +1,190 @@
+func NewInMemoryNetworkWithGateways(t *testing.T, cfg *network.InMemoryNetworkConfig) *inMemoryNetworkWithGateways {
+	t.Helper()
+
+	return &inMemoryNetworkWithGateways{
+		BaseInMemoryNetwork: *basenet.NewBaseInMemoryNetwork(
+			t, cfg, testkeyring.NewPreGeneratedAccountIterator(),
+		),
+	}
+}
+
+// DelegateAppToGateway delegates the application by address to the gateway by address.
+func (memnet *inMemoryNetworkWithGateways) DelegateAppToGateway(
+	t *testing.T,
+	appBech32 string,
+	gatewayBech32 string,
+) {
+	t.Helper()
+
+	args := []string{
+		gatewayBech32,
+		fmt.Sprintf("--%s=%s", flags.FlagFrom, appBech32),
+		fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
+		fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastSync),
+		fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(memnet.GetNetwork(t).Config.BondDenom, math.NewInt(10))).String()),
+	}
+	responseRaw, err := testcli.ExecTestCLICmd(memnet.GetClientCtx(t), cli.CmdDelegateToGateway(), args)
+	require.NoError(t, err)
+	var resp sdk.TxResponse
+	require.NoError(t, memnet.GetCosmosNetworkConfig(t).Codec.UnmarshalJSON(responseRaw.Bytes(), &resp))
+	require.NotNil(t, resp)
+	require.NotNil(t, resp.TxHash)
+	require.Equal(t, uint32(0), resp.Code)
+}
+
+// UndelegateAppFromGateway delegates the application by address from the gateway by address.
+func (memnet *inMemoryNetworkWithGateways) UndelegateAppFromGateway(
+	t *testing.T,
+	appBech32 string,
+	gatewayBech32 string,
+) {
+
+	args := []string{
+		gatewayBech32,
+		fmt.Sprintf("--%s=%s", flags.FlagFrom, appBech32),
+		fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
+		fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastSync),
+		fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(memnet.GetCosmosNetworkConfig(t).BondDenom, math.NewInt(10))).String()),
+	}
+	responseRaw, err := testcli.ExecTestCLICmd(memnet.GetClientCtx(t), cli.CmdUndelegateFromGateway(), args)
+	require.NoError(t, err)
+	var resp sdk.TxResponse
+	require.NoError(t, memnet.GetCosmosNetworkConfig(t).Codec.UnmarshalJSON(responseRaw.Bytes(), &resp))
+	require.NotNil(t, resp)
+	require.NotNil(t, resp.TxHash)
+	require.Equal(t, uint32(0), resp.Code)
+}
+
+// Start initializes the in-memory network and performs the following setup:
+//   - populates a new in-memory keyring with a sufficient number of pre-generated accounts.
+//   - configures the application module's genesis state using addresses corresponding
+//     to config.NumApplications number of the same pre-generated accounts which were
+//     added to the keyring.
+//   - configures the supplier module's genesis state using addresses corresponding to
+//     config.NumSuppliers number of the same pre-generated accounts which were added
+//     to the keyring.
+//   - creates the on-chain accounts in the accounts module which correspond to the
+//     pre-generated accounts which were added to the keyring.
```

Comment:
This little function does all that! Wow... 💁🏼‍♂️

Reply:
That's the single-responsibility principle at work! :muscle:

Reply:
Sounds SOLID